### PR TITLE
Add nth_algebric solver to the ode module

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

This is following on from a discussion in #6259 and goes some way to fixing that issue.

#### Brief description of what is fixed or changed

Adds a new ODE solver called `nth_algebraic`. The solver uses `sympy.solve` but teaches it to be able to invert derivatives algebraically. This means that ODEs that can be solved simply with algebra and integrals can be solved by `dsolve`.

This increases the number of ODEs that can be solved by `dsolve` and in some cases increases the number of solutions that `dsolve` can find. It does not work as well as it could though due to other parts of the `ode` module.

In particular `dsolve` (with `prep=True` (default)) preprocesses the ODE before passing it to the solver: [https://github.com/sympy/sympy/blob/sympy-1.3/sympy/solvers/deutils.py#L173](https://github.com/sympy/sympy/blob/sympy-1.3/sympy/solvers/deutils.py#L173)
This means that with this patch there are cases that `dsolve` can now solve provided the user passes `prep=False` but still cannot solve with the default `prep=True`.

For example, if `prep=True` then `_preprocess` will expand `Derivative(x*f(x), x, x, x)` using the product rule into `x*Derivative(f(x), (x, 3)) + 3*Derivative(f(x), (x, 2))`. This means that `nth_algebraic` still can't solve this ODE even though it could if it wasn't pre-expanded.

Also in some other cases `dsolve` simplifies equations removing solutions: [https://github.com/sympy/sympy/blob/sympy-1.3/sympy/solvers/ode.py#L1040](https://github.com/sympy/sympy/blob/sympy-1.3/sympy/solvers/ode.py#L1040)
For example in the case of `f(x)*(1+f(x).diff(x))` that line will divide by `f(x)` which removes the solution `f(x)=0` that would otherwise be found by `nth_algebraic`.

Because this solver can find algebraic solutions it can find some solutions that are *purely algebraic* and do not have integration constants e.g.:
```julia
In [4]: dsolve((f(x) - 1)*(f(x).diff(x) - 1))
Out[4]: [f(x) = 1, f(x) = C₁ + x]
```

In this example we have the solution `f(x)=1` which is purely algebraic and therefore doesn't have any integration constants.

In some cases it is possible that the additional algebraic solutions are redundant in the context of the general integral solutions since they represent a particular choice for the integration constants. I have had to add additional code to filter out solutions such as those so that only non-redundant solutions are returned:
```julia
In [5]: dsolve((f(x) - 1)*f(x).diff(x))
Out[5]: f(x) = C₁
```

The code for removing redundant solutions feels like it could be improved. At the moment it substitutes a load of values for `x` into both solutions and then uses `solve` to detect if the equations are inconsistent but there are problems with that (need to cll `.doit()` and don't know if the values can reasonably be substituted).

#### Other comments

As it stands I think that the solver in this PR represents an improvement in correctness for `dsolve`. It could be better if other parts of the code were changed.

I don't know what the best way is to do the redundancy checking. The question is: how can I tell that `f(x) = 1` is a special case of `f(x) = 1 + C1*x + C2*x**2`?

I also haven't checked what happens when initial conditions are combined with this solver. Since some of the returned solutions don't have integration constants it is possible that that would confuse the IC handling code.

It is also possible that the solver could be made more powerful if `solve` could be taught to understand other parts of the algebra of derivatives.

This results in a slow-down when running the test-suite (due to a slow-down in `dsolve` rather than the additional tests) of around 20%. The main performance cost is that the matching code calls `solve` to see if it can find solutions.

#### Release Notes

* solvers
  * Add an 'nth_algebraic' solver for ODEs that can be solved by algebra and integration

### CODE REVIEW

File: examples/notebooks/fibonacci.ipynb [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+      }
+     ], 
+     "prompt_number": 1
+    }, 
+    {
+     "cell_type": "markdown", 
+     "source": [
+      "And it does the trick. But that was kind of cheating - what if we don't know Fibonacci Sequence?", 
+      "", 
+      "##Solving the reccurence##",
```
Comment (jksuom) [inline examples/notebooks/fibonacci.ipynb]: reccurence -> recurrence

File: sympy/solvers/ode.py [hunk 2]
```diff
@@ -3968,6 +3980,168 @@ def _frobenius(n, m, p0, q0, p, q, x0, x, c, check=None):
...
+    solns = [Equality(func, soln) for soln in solns]
+    return {'var':var, 'solutions':solns}
+
+def ode_nth_algebraic(eq, func, order, match):
+    r"""
+    Solves an `n`\th order ordinary differential equation using algebra and
+    integrals.
+
+    There is no general form for the kind of equation that this can solve. The
+    the equation is solved algebraically treating differentiation as an
```
Comment (jksuom) [inline sympy/solvers/ode.py]: repeated 'the'

**Comment by oscarbenjamin** [comment]
Hi Kalevi,

That commit wasn't supposed to be In this PR. It should only change the ode
module and tests. I guess I'll need to open a new rebased PR.

**Comment by oscarbenjamin** [comment]
I'm closing this PR as it included irrelevant commits and I've opened #15255 with the commits I wanted to send in this PR.

**Pull request closed without merging**